feat(engine): sort items topologically

W95Psp opened the pull request:
- [x] add tests

karthikbhargavan reviewed (approved):
Much needed. Do we need to test this on all our code to be sure?

karthikbhargavan commented on engine/lib/dependencies.ml:
```diff
@@ -205,6 +206,15 @@ module Make (F : Features.T) = struct
+      List.find ~f:(ident_of >> Concrete_ident.equal name) items
+    in
+    ItemGraph.Topological.fold List.cons g []
+    (* |> List.rev *)
```
delete?

W95Psp replied:
Oh, right, thanks!

W95Psp commented:
I added an assert to make sure the reordering wasn't losing any item, turned out "orphan" items were discarded! I fixed that now.
So I thing merging this is fine.

W95Psp commented:
Ok, the CI is happy, I'm pressing merge
